Userland/Libraries/LibWeb/Bindings/WindowObject.cpp

```diff
@@ -437,4 +447,114 @@ JS_DEFINE_NATIVE_FUNCTION(WindowObject::inner_height_getter)
     return JS::Value(impl->inner_height());
 }
 
+JS_DEFINE_NATIVE_GETTER(WindowObject::scroll_x_getter)
+{
+    auto* impl = impl_from(vm, global_object);
+    if (!impl)
+        return {};
+    if (!impl->document().page())
+        return JS::Value(0.0);
+    return JS::Value(static_cast<double>(impl->document().page()->top_level_browsing_context().viewport_scroll_offset().x()));
+}
+
+JS_DEFINE_NATIVE_GETTER(WindowObject::scroll_y_getter)
+{
+    auto* impl = impl_from(vm, global_object);
+    if (!impl)
+        return {};
+    if (!impl->document().page())
+        return JS::Value(0.0);
+    return JS::Value(static_cast<double>(impl->document().page()->top_level_browsing_context().viewport_scroll_offset().y()));
+}
+
+enum class ScrollBehavior {
+    Auto,
+    Smooth
+};
+
+static void perform_a_scroll(Page& page, double x, double y, ScrollBehavior)
+{
+    // FIXME: Stop any existing smooth-scrolls
+    // FIXME: Implement smooth-scroll
+    page.client().page_did_request_scroll_to({ x, y });
+}
+
+// https://drafts.csswg.org/cssom-view/#dom-window-scroll
+JS_DEFINE_NATIVE_FUNCTION(WindowObject::scroll)
+{
+    auto* impl = impl_from(vm, global_object);
+    if (!impl)
+        return {};
+    if (!impl->document().page())
+        return {};
+    auto& page = *impl->document().page();
+
+    auto viewport_rect = page.top_level_browsing_context().viewport_rect();
+    JS::Value x_value = JS::Value(viewport_rect.x());
+    JS::Value y_value = JS::Value(viewport_rect.y());
+    String behavior_string = "auto";
+
+    if (vm.argument_count() == 1) {
+        auto& options = vm.argument(0).as_object();
+        if (auto left = options.get("left"); !left.is_undefined())
+            x_value = left;
+        if (auto top = options.get("top"); !top.is_undefined())
+            y_value = top;
+        behavior_string = options.get("behavior").to_string_without_side_effects();
+    } else if (vm.argument_count() == 2) {
+        x_value = vm.argument(0);
+        y_value = vm.argument(1);
+    }
+
+    double x = (x_value.is_finite_number() ? x_value.as_double() : 0);
+    double y = (y_value.is_finite_number() ? y_value.as_double() : 0);
+
+    // FIXME: Are we calculating the viewport in the way this function expects?
+    // FIXME: Handle overflow-directions other than top-left to bottom-right
+
+    ScrollBehavior behavior = (behavior_string == "smooth") ? ScrollBehavior::Smooth : ScrollBehavior::Auto;
+    perform_a_scroll(page, x, y, behavior);
+    return {};
```

Comment:
Returning empty `JS::Value` is probably not what you want since this is not something the user side of js should every get. 
I'm not sure but looking at the spec and the browsers I have I think it should just return `JS::js_undefined()`
Perhaps they mean https://heycam.github.io/webidl/#VoidFunction?

Reply:
Thanks, I made the assumption that `{}` is undefined. Fixed!